### DISCUSSION TITLE
Why does rootless nerdctl run two times of main()?

### DISCUSSION THREAD

**Discussion by ningziwen** [containerd/nerdctl#2013] [Q&A] [answered]
I changed one line ./cmd/nerdctl/main.go to add log for starting main.
```
func main() {
	logrus.Info("Start main")
	if err := xmain(); err != nil {
		HandleExitCoder(err)
		logrus.Fatal(err)
	}
}
```

Then if I run rootless nerdctl, it will print two times.
```
Admin:~/nerdctl (master) $ ./_output/nerdctl run  -it public.ecr.aws/amazonlinux/amazonlinux:2 bash
INFO[0000] Start main                                   
INFO[0000] Start main 
```

The rootful nerdctl only prints it for one time.

```
Admin:~/nerdctl (master) $ sudo ./_output/nerdctl run  -it public.ecr.aws/amazonlinux/amazonlinux:2 bash                                                                                                                 
INFO[0000] Start main
```

**Comment by AkihiroSuda** [comment, marked as answer]
To enter the namespaces.

https://github.com/containerd/nerdctl/blob/bb7bba5db37f00c723e5737f91b02ebb24494fc9/cmd/nerdctl/main.go#L229-L232

(Without this, nsentering to userns isn't possible with Go AFAICS due to threading)

**Comment by fahedouch** [comment]
```
	args := []string{
		"-r/",     // root dir (busybox nsenter wants this to be explicitly specified),
		"-w" + wd, // work dir
		"--preserve-credentials",
		"-m", "-n", "-U",
		"-t", strconv.Itoa(childPid),
		"-F", // no fork
```
@AkihiroSuda 

why are we entering to `mount`, `network` and  `user` ns but NOT `pid` namespace. 

Other question: how this entrering is handled in the case of rootfull mode ? ( you can answer me by providing me a link or simple phrase  ;) )

**Reply by AkihiroSuda** [reply]
> why are we entering to mount, network and user ns but NOT pid namespace.

Because PID namespace is not unshared.
UserNS, MountNS, and NetNS are unshared to gain the fake privileges inside it.

> Other question: how this entrering is handled in the case of rootfull mode ? ( you can answer me by providing me a link or simple phrase ;) )

Entering is not needed for rootful because no namespace is unshared.

**Reply by fahedouch** [reply]
> Because PID namespace is not unshared.
UserNS, MountNS, and NetNS are unshared to gain the fake privileges inside it.

got the point, the idea is to make a fake user/net ns for the rootless user 

> Entering is not needed for rootful because no namespace is unshared.

So are they cloned or forked to ensure isolation ?

**Reply by ningziwen** [reply]
```
// reexec /proc/self/exe with `nsenter` into RootlessKit namespaces 
```
This comment makes sense to me. Executing self should go back to main.

But based on the [args](https://github.com/containerd/nerdctl/blob/main/pkg/rootlessutil/parent.go#L91) of `nsenter`, I didn't find the logic of executing /proc/self/exe? Did I miss anything?

**Reply by AkihiroSuda** [reply]
Looks like we are just using `os.Args[0]` as the self exe

https://github.com/containerd/nerdctl/blob/2e5cdacc3e66ce4ddb1c35dd132d81816d784834/pkg/rootlessutil/parent.go#L99
